packages/labs/analyzer/src/lib/javascript/mixins.ts

```diff
@@ -0,0 +1,205 @@
+  // a class expression, hence why we effectively disallow them here. However,
+  // we may want to loosen this constraint for JS users some day
+  if (!analyzer.typescript.isBlock(functionBody)) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to have a block function body; arrow-function class ` +
+        `expression syntax is not supported.`,
+      analyzer
+    );
+    return undefined;
+  }
+  let classDeclaration!: ts.ClassDeclaration;
+  let returnStatement!: ts.ReturnStatement;
+  functionBody.statements.forEach((s) => {
+    if (analyzer.typescript.isClassDeclaration(s)) {
+      classDeclaration = s;
+    }
+    if (analyzer.typescript.isReturnStatement(s)) {
+      returnStatement = s;
+    }
+  });
+  if (!classDeclaration) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain a class declaration statement.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (!returnStatement) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain a return statement returning a class.`,
+      analyzer
+    );
+    return undefined;
+  }
+  const extendsClause = classDeclaration.heritageClauses?.find(
+    (c) => c.token === analyzer.typescript.SyntaxKind.ExtendsKeyword
+  );
+  if (extendsClause === undefined) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain class declaration extending a superClass argument to function.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (extendsClause.types.length !== 1) {
+    analyzer.addDiagnostic(
+      createDiagnostic({
+        typescript: analyzer.typescript,
+        node: extendsClause,
+        message:
+          'Internal error: did not expect a mixin class extends clause to have multiple types',
+        code: DiagnosticCode.UNSUPPORTED,
+        category: analyzer.typescript.DiagnosticCategory.Warning,
+      })
+    );
+    return undefined;
+  }
+  const superClassArgIdx = findSuperClassArgIndexFromHeritage(
+    possibleSuperClasses,
+    extendsClause.types[0].expression,
+    analyzer
+  );
+  if (superClassArgIdx < 0) {
+    analyzer.addDiagnostic(
+      createDiagnostic({
+        typescript: analyzer.typescript,
+        node: extendsClause,
+        message:
+          'Did not find a "superClass" argument used in the extends clause of mixin class.',
+        code: DiagnosticCode.UNSUPPORTED,
+        category: analyzer.typescript.DiagnosticCategory.Warning,
+      })
+    );
+    return undefined;
+  }
+
+  const classDeclarationName = classDeclaration.name?.text ?? name;
+
+  return {
+    node: fn,
+    name,
+    superClassArgIdx,
+    classDeclaration: getClassDeclaration(
+      classDeclaration,
+      classDeclarationName,
+      analyzer,
+      undefined,
+      true /* isMixinClass */
+    ),
+  };
+};
+
+const findSuperClassArgIndexFromHeritage = (
```

Comment:
Documentation on this function would be helpful since it's quite impressive what this function is doing, and it took me a while to understand it.

If I understand correctly, this function is finding which argument is the superClass.

Reply:
I may be completely misunderstanding this function as well, but it may be worth adding an additional test case where a mixin applies some other mixin.

Reply:
you understood correctly i think

it'll basically recursively resolve your example:

- initially it'll have the expression `someOtherMixin(someArg, superClass)` and some possible names of `['superClass', 'someArg']`
- it'll then check if `someArg` is one of the possible names
- it is so it'll return index `1` (because its `[1]` of the possible names)

so you're right, it'll pick up `someArg` as the class arg rather than `superClass`

not sure what we can do about that though 🤔

Reply:
Good catch here, this does seem wrong. I think in the `isCallExpression` branch, rather than recursing into `findSuperClassArgIndexFromHeritage()`, this function likely needs to try to look up / dereference the mixin that's being called and return the index of its own argument passed into the mixin's superclass argument; that will recurse until a mixin extends one of its arguments directly (iow, taking the `isIdentifier` branch) rather than calling another mixin.

So in the example, we'd look up `someOtherMixin`, it would have a `superClassArgIdx` of `1`, and so we'd find the index of the `superClass` argument, and search the current mixin's arg list for that name and return `0`.

As Justin said in the other comment, the name searches could be likely be done with symbol declaration identity, but that could be a separate change.

Reply:
I think we might need to add something to the custom elements manifest here to identify which parameter of a mixin is the superclass.

We can do something internally to the analyzer in the mean time by just adding that to the model object.